### ISSUE TITLE
Feature: Option to randomize wallpapers every 5 minutes

### ISSUE THREAD

**Issue by chrissolanilla** [anufrievroman/waypaper#69]
Hi, I really like the easy of use that waypaper provides. One thing that I usually have when it comes to wall papers is for it to randomly change to a different image after a certain amount of time. Usually I achieve this with a script. When it comes to other operating systems that are not Linux, they support this out of the box sometimes. I would like to propose this idea, and will try to fork my own copy of this repository but with an additional button or option to randomize the wallpaper after some time like 1 minute, 5 minutes, 30 minutes, 1 hour.

**Comment by anufrievroman** [comment]
Hi, thank you for the suggestion. One issue with that is to do that we need to have a running process, i.e. either the GUI should keep running or we need a daemon that will keep counting the time (which is essentially the same as just your own script). Currently, I am not really convinced that we need to convert `waypaper` into a daemon-like program just for this one feature. Maybe documentation with example of a script to achieve this would be enough.

**Comment by chrissolanilla** [comment]
Thank you for telling me a simple way to do the bash script. I was thinking maybe the program could create a bash script for the user, but I understand why you may be adverse to the idea of this feature.

**Comment by chrissolanilla** [comment]
I have finished my fork of this project and have opened a pull request. The details are in the pull request. It would mean a lot to me if you accepted the pull request, thanks.

**Comment by anufrievroman** [comment]
Hi, it's been a while. I actually have another idea how to implement randomization on set time without creating more daemons or this kind of things. There is a backend called [wpaperd](https://github.com/danyspin97/wpaperd) which supports an option of changing wallpaper every set amount of time. So, we could just implement this backend and control this `duraiton` parameter in the UI of waypaper. 

Also, now we moved quite far in the code base, so I'll need to decline your pull request as it cannot be merged anymore. But I think we will implement this functionality in the next version.

**Comment by jgregoire** [comment]
Have you considered setting up an external timer with cron or systemd?

Linux has these features for just such a task. :)

**Comment by anufrievroman** [comment]
@jgregoire that's might be a good idea. Assuming that we want this to be a functionality of the app (not external script), we need the ability to run a command `waypaper --random` even N minutes, and we need the ability to both start/stop it from the GUI with a button and automatically start it with `waypaper --restore` (i.e. when it's launched at the start of the system). How would you do this with cron?

**Comment by jgregoire** [comment]
Been too long since I've used a distro without systemd, but IIRC the GUI would have to add an entry to crontab, and delete or comment it out to disable it.

Systemd is easy. Just provide a timer file and a service file to the user. I can write those up and test if that's helpful.

**Comment by anufrievroman** [comment]
Yes, could you provide some examples how to do it? Maybe we can just add it into documentation and that's enough

**Comment by jgregoire** [comment]
`~/.config/systemd/user/waypaper.timer`
```service
[Unit]
Description=Set a random wallpaper every minute

[Timer]
Persistent=true
OnCalendar=*:0/1

[Install]
WantedBy=timers.target
```
`~/.config/systemd/user/waypaper.service`
```service
[Unit]
Description=Set a random wallpaper with waypaper

[Service]
Type=oneshot
ExecStart=/home/james/.local/bin/waypaper --random
```

To test, run `$ systemctl --user start waypaper.timer`
To make persistent, run `$ systemctl --user enable waypaper.timer`

**Comment by anufrievroman** [comment]
For the record, I implemented the auto-change of the wallpaper via native mpvpaper function, but to be honest it looks quite awful, the change is glitchy, and the order is not random. So I'll probably get rid of that and make our own deamon or just forget about it.

**Comment by anufrievroman** [comment]
Okay, I added the instructions by @jgregoire to the [Documentation](https://anufrievroman.gitbook.io/waypaper/usage), so I think I'll close this issue for now. Thank you everyone for the discussions!

**Comment by khaosdoctor** [comment]
Just adding to this. I created the units but unfortunately it didn't work. The wallpaper will blink and go back to gray again, it will only work if the unit type is `exec` because it keeps the process running, but then the timer unit will not restart the service unit and nothing happens.

I tried

```ini
[Unit]
Description=Change wallpaper with Waypaper and update colors

[Service]
Type=exec
Environment=DISPLAY=:0
Environment=WAYLAND_DISPLAY=wayland-1
Environment=XDG_CURRENT_DESKTOP=Hyprland
Environment=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
ExecStart=waypaper --random
```

Then it didn't work, but there's a way if you keep the process running with a subshell:

```ini
[Unit]
Description=Change wallpaper with Waypaper and update colors

[Service]
Type=exec
Environment=DISPLAY=:0
Environment=WAYLAND_DISPLAY=wayland-1
Environment=XDG_CURRENT_DESKTOP=Hyprland
Environment=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
ExecStart=/bin/bash -c 'w=$(waypaper --random)'
```

Then it works, but the timer won't pick it up.

**Comment by bruno-unna** [comment]
@khaosdoctor 

I faced the very same issue. I think I found a way to fix it: use the option `KillMode` set to `process` in the service file:

```
[Unit]
Description=Set a random wallpaper with waypaper

[Service]
Type=oneshot
ExecStart=waypaper --random
KillMode=process
``` 

Hope that helps.
